File: ObjcFrame/ObjcFrame/Base/Controller/BaseNavigationController.m
//
//  BaseNavigationController.m
//  MengTianXIa
//
//  Created by zl on 2019/8/9.
//  Copyright © 2019 qttx. All rights reserved.
//

#import "BaseNavigationController.h"

@interface BaseNavigationController () <UIGestureRecognizerDelegate>

@end

@implementation BaseNavigationController

+ (void)initialize {
    UINavigationBar *bar = [UINavigationBar appearance];
    // 去除导航栏的毛玻璃效果
    bar.translucent = NO;
    [bar setBackgroundImage:[UIImage new]  forBarPosition:UIBarPositionAny barMetrics:UIBarMetricsDefault];
    // 隐藏导航栏底部分割线
    [bar setShadowImage:[UIImage new]];
    // 设置标题文字
    NSMutableDictionary *attrs = [NSMutableDictionary dictionary];
    attrs[NSFontAttributeName] = [UIFont boldSystemFontOfSize:18];
    [bar setTitleTextAttributes:attrs];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // 设置手势代理,开启向右滑返回功能
    self.interactivePopGestureRecognizer.delegate = self;
}


/**
 重写push方法实现页面跳转后的各种设置
 
 @param viewController 目的控制器
 @param animated 是否a带有动画效果
 */
- (void)pushViewController:(UIViewController *)viewController animated:(BOOL)animated {
    // 子页面隐藏底部TabBar按钮
    if (self.childViewControllers.count > 0) {
        viewController.hidesBottomBarWhenPushed = YES;
        
        // 自定义子页面的返回按钮样式
        UIButton *backBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [backBtn setImage:[UIImage imageNamed:@"nav_back_btn_n"] forState:UIControlStateNormal];
        [backBtn setImage:[UIImage imageNamed:@"nav_back_btn_n"] forState:UIControlStateHighlighted];
        //        [backBtn setTitle:@"返回" forState:UIControlStateNormal];
        //        [backBtn setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
        //        [backBtn setTitleColor:MYColor(251, 32, 37) forState:UIControlStateHighlighted];
        //        backBtn.backgroundColor = [UIColor redColor];
        backBtn.contentEdgeInsets = UIEdgeInsetsMake(10, 0, 10, 30);
        [backBtn sizeToFit];
        [backBtn addTarget:self action:@selector(goBack) forControlEvents:UIControlEventTouchUpInside];
        
        viewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:backBtn];
    }
    
    [super pushViewController:viewController animated:animated];
}

- (void)goBack {
    [self popViewControllerAnimated:YES];
}

#pragma mark - UIGestureRecognizerDelegate
// 启用手势返回
- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer {
    return self.childViewControllers.count > 1;
}
@end

File: SwiftFrame-1.0/MengTianXia/Base/BaseNavigationController.swift
//
//  MyNavigationController.swift
//  swift-APP框架搭建
//
//  Created by zl on 2019/5/30.
//  Copyright © 2019 ZHUNJIEE. All rights reserved.
//

import UIKit

class BaseNavigationController: UINavigationController, UIGestureRecognizerDelegate {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setUpNavigationBar()
        
        // 启用左滑返回手势
        interactivePopGestureRecognizer?.delegate = self
    }
    
    override func pushViewController(_ viewController: UIViewController, animated: Bool) {
        if children.count > 0 {
            // 跳转后隐藏底部tabbar
            viewController.hidesBottomBarWhenPushed = true
            
            // 自定义返回按钮样式
            let backBtn = UIButton(type: .custom)
            backBtn.setImage(UIImage(named: "nav_back_btn_n"), for: .normal)
            backBtn.setImage(UIImage(named: "nav_back_btn_n"), for: .highlighted)
            backBtn.setTitle("返回", for: .normal)
            backBtn.setTitleColor(UIColor.orange, for: .normal)
            backBtn.contentEdgeInsets = UIEdgeInsets(top: 10, left: 0, bottom: 10, right: 10)
            backBtn.sizeToFit()
            backBtn.addTarget(self, action: #selector(goBack), for: .touchUpInside)
            viewController.navigationItem.leftBarButtonItem = UIBarButtonItem(customView: backBtn)
        }
        super.pushViewController(viewController, animated: animated)
    }
    
    
    /// 启用左滑返回手势
    func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
        return children.count > 1
    }
}

// MARK: UI界面
extension BaseNavigationController {
    fileprivate func setUpNavigationBar() {
        let navBar = UINavigationBar.appearance()
        // 取消毛玻璃效果
        navBar.isTranslucent = false
        // 去除分割线
        navBar.shadowImage = UIImage()
        navBar.setBackgroundImage(UIImage(), for: .default)
        // 设置标题文字属性
        var attrs = [NSAttributedString.Key : Any]()
        attrs[NSAttributedString.Key.font] = UIFont.systemFont(ofSize: 18.0)
        navBar.titleTextAttributes = attrs
    }
}

// MARK: 事件监听
extension BaseNavigationController {
    /// 返回前一个界面
    @objc func goBack() {
        popViewController(animated: true)
    }
}
